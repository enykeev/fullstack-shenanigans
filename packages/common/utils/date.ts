const TIME_UNITS = {
  second: 1,
  minute: 60,
  hour: 60 * 60,
  day: 24 * 60 * 60,
  week: 7 * 24 * 60 * 60,
  // After that you need more context a single number can provide. So we'll just skip that for now.
};

const SORTED_TIME_UNIT_PAIRS = [...Object.entries(TIME_UNITS)].sort(
  (a, b) => a[1] - b[1],
);

export function humanTime(seconds: number) {
  let index = SORTED_TIME_UNIT_PAIRS.length;
  for (let i = 1; i < SORTED_TIME_UNIT_PAIRS.length; i++) {
    if (seconds < SORTED_TIME_UNIT_PAIRS[i][1]) {
      index = i;
      break;
    }
  }
  const [unit, unitValue] = SORTED_TIME_UNIT_PAIRS[index - 1];
  const value = seconds / unitValue;
  if (value === 1) {
    return {
      value: value,
      units: unit,
    };
  } else {
    // TODO: replace with proper i18n
    return {
      value: value,
      units: `${unit}s`,
    };
  }
}
