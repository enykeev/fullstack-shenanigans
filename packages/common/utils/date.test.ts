import { describe, expect, test } from "bun:test";
import { humanTime } from "./date";

describe("Date Util", () => {
  test("humanTime", () => {
    expect(humanTime(0)).toMatchSnapshot();
    expect(humanTime(0.1)).toMatchSnapshot();
    expect(humanTime(1)).toMatchSnapshot();
    expect(humanTime(60 - 1)).toMatchSnapshot();
    expect(humanTime(60)).toMatchSnapshot();
    expect(humanTime(1.5 * 60)).toMatchSnapshot();
    expect(humanTime(60 * 60 - 1)).toMatchSnapshot();
    expect(humanTime(60 * 60)).toMatchSnapshot();
    expect(humanTime(24 * 60 * 60 - 1)).toMatchSnapshot();
    expect(humanTime(24 * 60 * 60)).toMatchSnapshot();
    expect(humanTime(7 * 24 * 60 * 60 - 1)).toMatchSnapshot();
    expect(humanTime(7 * 24 * 60 * 60)).toMatchSnapshot();
    expect(humanTime(30 * 24 * 60 * 60 - 1)).toMatchSnapshot();
    expect(humanTime(30 * 24 * 60 * 60)).toMatchSnapshot();
    expect(humanTime(52 * 7 * 24 * 60 * 60)).toMatchSnapshot();
  });
});
