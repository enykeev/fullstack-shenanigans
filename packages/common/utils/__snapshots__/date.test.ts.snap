// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Date Util humanTime 1`] = `
{
  "units": "seconds",
  "value": 0,
}
`;

exports[`Date Util humanTime 2`] = `
{
  "units": "seconds",
  "value": 0.1,
}
`;

exports[`Date Util humanTime 3`] = `
{
  "units": "second",
  "value": 1,
}
`;

exports[`Date Util humanTime 4`] = `
{
  "units": "seconds",
  "value": 59,
}
`;

exports[`Date Util humanTime 5`] = `
{
  "units": "minute",
  "value": 1,
}
`;

exports[`Date Util humanTime 6`] = `
{
  "units": "minutes",
  "value": 1.5,
}
`;

exports[`Date Util humanTime 7`] = `
{
  "units": "minutes",
  "value": 59.983333333333334,
}
`;

exports[`Date Util humanTime 8`] = `
{
  "units": "hour",
  "value": 1,
}
`;

exports[`Date Util humanTime 9`] = `
{
  "units": "hours",
  "value": 23.99972222222222,
}
`;

exports[`Date Util humanTime 10`] = `
{
  "units": "day",
  "value": 1,
}
`;

exports[`Date Util humanTime 11`] = `
{
  "units": "days",
  "value": 6.999988425925926,
}
`;

exports[`Date Util humanTime 12`] = `
{
  "units": "week",
  "value": 1,
}
`;

exports[`Date Util humanTime 13`] = `
{
  "units": "weeks",
  "value": 4.285712632275132,
}
`;

exports[`Date Util humanTime 14`] = `
{
  "units": "weeks",
  "value": 4.285714285714286,
}
`;

exports[`Date Util humanTime 15`] = `
{
  "units": "weeks",
  "value": 52,
}
`;
